### PR TITLE
fix(core): fix possible database crash on Windows when disk runs out of space

### LINKED ISSUE #2823: 6.6.1 version is not stable on windows os
### Describe the bug

 
the version 6.61  is not stable ,the status of qdb frequently be inactive ,there are two crash logs,
 my questdb is on windows11 os
[crash+1.log](https://github.com/questdb/questdb/files/10101718/crash%2B1.log)
[crash+0.log](https://github.com/questdb/questdb/files/10101719/crash%2B0.log)

### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bluestreak01**
fixes #2823

Fixes crash by removing race condition allocating disk space on the cusp of space running out. Added test to prove the fix.

We used to rely on `mmap()` call to allocate disk space on windows. This cannot be done reliably however when two or more `mmap()` calls are racing. For that reason we cannot bypass `allocate()` call as it guarantees multi-threaded stability.

Also in this PR `Files.getDiskSize()` is added  to ensure the condition of overallocating disk can be simulated reliably in a test. In that test I noticed that `allocate()` calls on OSX are painfully slow. This seems to be a fault of the OS as there is no other way to pre-allocate a file, other than already employed by QuestDB.

### Refactoring

This PR awards OSX its own `files.c` source code, which is that of `FreeBSD`. I got rid of systematic

```
#ifdef __APPLE_
```

usages, in favour of CMake using different file for OSX

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 7 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Files.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 7 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Files.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|2|100.00%|

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 7 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Files.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|2|100.00%|

**Review by amunra — CHANGES_REQUESTED** [review]
* Some minor code clarity changes (optional)
* Missing error handling should be implemented.

**Comment by amunra** [comment]
What would also be nice is to include logic to log the actual error in case `EXCEPTION_IN_PAGE_ERROR` happens again for a different reason.
See: https://devblogs.microsoft.com/oldnewthing/20120608-00/?p=7423

File: core/src/main/c/freebsd/files.c [hunk 1]
```diff
@@ -126,21 +131,55 @@ size_t copyData0(int inFd, int outFd, off_t fromOffset, jlong length) {
...
+        uint64_t k = buf.f_bavail * buf.f_bsize;
+        return (jlong) k;
+    }
+    return -1;
+}
+
+JNIEXPORT jboolean JNICALL Java_io_questdb_std_Files_setLastModified
+        (JNIEnv *e, jclass cl, jlong lpszName, jlong millis) {
+    struct timeval t[2];
+    gettimeofday(t, NULL);
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/c/freebsd/files.c]: good tip, thank you 👍 

fixed in https://github.com/questdb/questdb/pull/2827/commits/c29a957f3a4fd129da4addc65122a22e6598197c

File: core/src/main/c/linux/files.c [hunk 2]
```diff
@@ -368,3 +370,52 @@ JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_getFileSystemStatus
     }
     return 0;
 }
+
+JNIEXPORT jboolean JNICALL Java_io_questdb_std_Files_setLastModified
+        (JNIEnv *e, jclass cl, jlong lpszName, jlong millis) {
+    struct timeval t[2];
+    gettimeofday(t, NULL);
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/c/linux/files.c]: fixed in https://github.com/questdb/questdb/pull/2827/commits/c29a957f3a4fd129da4addc65122a22e6598197c

File: core/src/main/c/linux/files.c [hunk 3]
```diff
@@ -368,3 +370,52 @@ JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_getFileSystemStatus
...
+    t[1].tv_sec = millis / 1000;
+    t[1].tv_usec = ((millis % 1000) * 1000);
+    return (jboolean) (utimes((const char *) lpszName, t) == 0);
+}
+
+JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_getDiskSize(JNIEnv *e, jclass cl, jlong lpszPath) {
+    struct statvfs buf;
+    if (statvfs((const char *) lpszPath, &buf) == 0) {
+        unsigned long long k = ((unsigned long long) buf.f_bavail) * buf.f_bsize;
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/c/linux/files.c]: this seems to work on linux 👍

File: core/src/main/c/windows/files.c [hunk 4]
```diff
@@ -884,3 +884,24 @@ JNIEXPORT jint JNICALL Java_io_questdb_std_Files_rename(JNIEnv *e, jclass cl, jl
     SaveLastError();
     return ERROR_NOT_SAME_DEVICE == GetLastError() ? FILES_RENAME_ERR_EXDEV : FILES_RENAME_ERR_OTHER;
 }
+
+JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_getDiskSize(JNIEnv *e, jclass cl, jlong lpszPath) {
+    ULARGE_INTEGER lpFreeBytesAvailableToCaller;
+    if ((void *) lpszPath != NULL) {
+        int len = MultiByteToWideChar(CP_UTF8, 0, (LPCCH) lpszPath, -1, NULL, 0);
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/c/windows/files.c]: done in https://github.com/questdb/questdb/pull/2827/commits/c29a957f3a4fd129da4addc65122a22e6598197c

Also applied to all usages of `MultiByteToWideChar()`

File: core/src/main/c/windows/files.c [hunk 5]
```diff
@@ -884,3 +884,24 @@ JNIEXPORT jint JNICALL Java_io_questdb_std_Files_rename(JNIEnv *e, jclass cl, jl
...
+        if (len > 0) {
+            wchar_t buf[len];
+            MultiByteToWideChar(CP_UTF8, 0, (LPCCH) lpszPath, -1, buf, len);
+            if (GetDiskFreeSpaceExW(buf, &lpFreeBytesAvailableToCaller, NULL, NULL)) {
+                return (jlong) lpFreeBytesAvailableToCaller.QuadPart;
+            }
+            SaveLastError();
+        }
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/c/windows/files.c:898]: good spot, there was missing errno capture, fixed in https://github.com/questdb/questdb/pull/2827/commits/c29a957f3a4fd129da4addc65122a22e6598197c

File: core/src/test/java/io/questdb/griffin/ColumnPurgeJobTest.java [hunk 6]
```diff
@@ -170,7 +170,7 @@ public void testPurge() throws Exception {
 
     @Test
     public void testPurgeCannotAllocateFailure() throws Exception {
-        int deadline = Os.type == Os.WINDOWS ? 152 : 105;
+        int deadline = Os.type == Os.WINDOWS ? 92 : 105;
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/test/java/io/questdb/griffin/ColumnPurgeJobTest.java:173]: on different platforms java calls different OS functions with different frequency in each test, this is just a nature of the beast. The change aims at failing the test in the same place as the code on master. The magic number `92` just happened to be the value of the counter in the place we want the fault to occur.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 7 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Files.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 7 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Files.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|2|100.00%|

**Review by amunra — APPROVED** [review]
Looking good!

**Pull request merged**